check_limits.py

```diff
@@ -1,18 +1,38 @@
+from report_vitals import report_normal_vitals,report_abnormal_vitals
 
-def battery_is_ok(temperature, soc, charge_rate):
-  if temperature < 0 or temperature > 45:
-    print('Temperature is out of range!')
-    return False
-  elif soc < 20 or soc > 80:
-    print('State of Charge is out of range!')
-    return False
-  elif charge_rate > 0.8:
-    print('Charge rate is out of range!')
-    return False
+bms_attribute_limits = {
+     'TEMPERATURE': {'min': 0, 'max': 45},
+     'SOC': {'min': 20, 'max': 80},
+     'CHARGE_RATE': {'min': 0,'max': 0.8}
+        }
 
-  return True
 
+def get_battery_vitals(bms_attribute):
+    offLimits = []
+    for bms_name,bms_value in bms_attribute.items():        
+        get_offLimitVitals(bms_name,bms_value,offLimits)       
+    return offLimits
 
+def get_offLimitVitals(bms_name,bms_value,offLimits):
+    low_breach_val = bms_attribute_limits[bms_name.upper()]['min']
+    high_breach_val = bms_attribute_limits[bms_name.upper()]['max']
+    if (bms_value < low_breach_val):
+        offLimits.append(bms_name.capitalize()+" Low Breach: "+str(low_breach_val))
+    elif (bms_value > high_breach_val):
+        offLimits.append(bms_name.capitalize()+" High Breach: "+str(high_breach_val))
+
+def battery_is_ok(bms_attribute):
+    bms_vitals = get_battery_vitals(bms_attribute)
+    if len(bms_vitals) == 0:
+        report_normal_vitals()
+        return True
+    else :
+        report_abnormal_vitals(bms_vitals)
+        return False
+    
 if __name__ == '__main__':
-  assert(battery_is_ok(25, 70, 0.7) is True)
-  assert(battery_is_ok(50, 85, 0) is False)
+  assert(battery_is_ok({'temperature': 25,'Soc': 70, 'Charge_rate': 0.7}) is True)  #all attributes fine
```

Comment:
Nice addition of tests. Any reason why upper limit-breaches are only tested? In general, how do you know that you have covered all cases?

Reply:
![image](https://user-images.githubusercontent.com/77484700/110238297-ec226180-7f66-11eb-9bae-cb517c345080.png)
I am checking the lower breach as well.